### ISSUE TITLE
[pyclustering.cluster.kmedians] exception: access violation reading 0x(Memory Address)

### ISSUE THREAD

**Issue by jordiarjona** [annoviko/pyclustering#428]
Hi,

When using kmedians, i get an error related to an access violation when reading a memory address. This happens if i use ccore=True. If i use ccore=False kmedians_obj.process() returns no medians or clusters. My guess is that it is related to the number of clusters (and maybe tolerance), although i am not sure. It happens when the ratio number of clusters to points is below 1/10 (The number of values and clusters i was using when got it was between 380-450 for 45 clusters). However, it might be interesting to try to capture the error so it is more informative.

Thanks a lot for your work!

**Comment by annoviko** [comment]
Hello, @jordiarjona ,

Thank you for problem reporting! May I ask you to share data or data example and used code for processing, or at least parameters of the algorithm? And could you please provide version of the pyclustering library? It helps to resolve the problem and prevent it appearance in further versions.

**Comment by jordiarjona** [comment]
Sure!
Pyclustering version is 0.8, running on windows and using cython 0.28.1 (in case the latter is relevant).
Here you have an example of data for which it failed. There are 382 pairs of values and i ask for 46 clusters. Values are coordinates and the number of clusters depends on the distance covered.

I choose the clusters using:
cluster_ids = random.sample(range(0, len(data)), num_clusters)
where data is, for isntance, the array provided below.

Then i run it like:
kmedians_obj = kmedians(data, initial_centers, tolerance=tolerance, ccore=True)
kmedians_obj.process()

I have tried different values for tolerance, in the range of 1-10.

Hope it helps! Let me know if you require something else.

[[59.00732, 9.748167], [59.00608, 9.749117], [59.0047, 9.749933], [59.00197, 9.752466], [58.99683, 9.753333], [58.99092, 9.7548], [58.98293, 9.761434], [58.9711, 9.774533], [58.95928, 9.78755], [58.95181, 9.795783], [58.94381, 9.804183], [58.93625, 9.81235], [58.9289, 9.820683], [58.92143, 9.8287], [58.91057, 9.8403], [58.90255, 9.849017], [58.89521, 9.856934], [58.8869, 9.866067], [58.87553, 9.878667], [58.86497, 9.889983], [58.8573, 9.89845], [58.84959, 9.9069], [58.84223, 9.914884], [58.83481, 9.92315], [58.82695, 9.93185], [58.81477, 9.945217], [58.8069, 9.953484], [58.79867, 9.962133], [58.79059, 9.97115], [58.77859, 9.983717], [58.77103, 9.9918], [58.76355, 9.999833], [58.75721, 10.0067], [58.74893, 10.0158], [58.74047, 10.02515], [58.72737, 10.03937], [58.7188, 10.04872], [58.70957, 10.05867], [58.70957, 10.05867], [58.7003, 10.06865], [58.69087, 10.07905], [58.68242, 10.08802], [58.67041, 10.10088], [58.66212, 10.10995], [58.64967, 10.12343], [58.6417, 10.13223], [58.63377, 10.14108], [58.62527, 10.14988], [58.61752, 10.15813], [58.60868, 10.16748], [58.59981, 10.17708], [58.59137, 10.18638], [58.58263, 10.19573], [58.57388, 10.20518], [58.56557, 10.21418], [58.5574, 10.22297], [58.54913, 10.23178], [58.54062, 10.24105], [58.53297, 10.24932], [58.52147, 10.26177], [58.51273, 10.27118], [58.50319, 10.28158], [58.4944, 10.29077], [58.48701, 10.29885], [58.47888, 10.30772], [58.4704, 10.31668], [58.46295, 10.32482], [58.45445, 10.33397], [58.4463, 10.3426], [58.43765, 10.35173], [58.42501, 10.36547], [58.41565, 10.37557], [58.40285, 10.38947], [58.38992, 10.40322], [58.37547, 10.41855], [58.32952, 10.46788], [58.32021, 10.47767], [58.30981, 10.4889], [58.30165, 10.49767], [58.29315, 10.50665], [58.27977, 10.52113], [58.27192, 10.52938], [58.2639, 10.53797], [58.2551, 10.54743], [58.2468, 10.55637], [58.23839, 10.56515], [58.2292, 10.57483], [58.22025, 10.58462], [58.21157, 10.59408], [58.20295, 10.60327], [58.19456, 10.61202], [58.18628, 10.6209], [58.17787, 10.62978], [58.16988, 10.63838], [58.16088, 10.64808], [58.14777, 10.66178], [58.13787, 10.67243], [58.1287, 10.68223], [58.12001, 10.69155], [58.1123, 10.69968], [58.10327, 10.70945], [58.0954, 10.71788], [58.08748, 10.72602], [58.07924, 10.7346], [58.06745, 10.74697], [58.05943, 10.75555], [58.05093, 10.7647], [58.04268, 10.7735], [58.03439, 10.78258], [58.02615, 10.79122], [58.01826, 10.79952], [58.00659, 10.81205], [57.9992, 10.82007], [57.98999, 10.82982], [57.97752, 10.84287], [57.96968, 10.85125], [57.96248, 10.85882], [57.95413, 10.86762], [57.94181, 10.88073], [57.9338, 10.88932], [57.92579, 10.89768], [57.91763, 10.90615], [57.90585, 10.91867], [57.89428, 10.93085], [57.88527, 10.94045], [57.87368, 10.953], [57.86172, 10.96552], [57.85379, 10.97377], [57.84621, 10.98168], [57.8386, 10.98962], [57.83064, 10.99805], [57.82173, 11.00818], [57.81068, 11.02028], [57.80255, 11.0294], [57.79147, 11.04185], [57.77887, 11.05575], [57.77141, 11.06407], [57.76348, 11.07295], [57.75505, 11.08228], [57.74653, 11.0918], [57.73868, 11.1004], [57.73081, 11.10922], [57.71907, 11.12218], [57.71121, 11.13087], [57.70299, 11.14005], [57.69519, 11.14865], [57.6826, 11.16263], [57.67375, 11.1727], [57.66193, 11.18583], [57.65295, 11.19558], [57.64387, 11.20555], [57.6356, 11.21435], [57.6281, 11.223], [57.6206, 11.23148], [57.61265, 11.24007], [57.60043, 11.25343], [57.59115, 11.26402], [57.57895, 11.27763], [57.57141, 11.28597], [57.56385, 11.29423], [57.5556, 11.30322], [57.54568, 11.3143], [57.53828, 11.32232], [57.52979, 11.33168], [57.51763, 11.34508], [57.51012, 11.35345], [57.50108, 11.36333], [57.49237, 11.3731], [57.48067, 11.38583], [57.47219, 11.3953], [57.4639, 11.40435], [57.45173, 11.41365], [57.44187, 11.42048], [57.43205, 11.42717], [57.41947, 11.43558], [57.41167, 11.441], [57.40077, 11.4483], [57.39212, 11.45422], [57.38327, 11.46003], [57.37401, 11.4664], [57.3653, 11.47247], [57.35732, 11.47797], [57.34832, 11.48438], [57.33887, 11.49038], [57.32941, 11.49658], [57.3211, 11.50253], [57.30707, 11.51188], [57.29687, 11.51885], [57.28647, 11.52617], [57.27674, 11.53253], [57.267, 11.53918], [57.25739, 11.54578], [57.24457, 11.5546], [57.23606, 11.56048], [57.22372, 11.56877], [57.21139, 11.57723], [57.20312, 11.58292], [57.19412, 11.58897], [57.18115, 11.59787], [57.173, 11.60348], [57.16087, 11.61147], [57.15123, 11.61765], [57.14281, 11.62333], [57.13186, 11.63032], [57.0633, 11.67378], [57.05493, 11.67932], [57.04108, 11.68862], [57.01492, 11.70522], [57.00315, 11.71283], [56.9586, 11.74217], [56.942, 11.75378], [56.90321, 11.78072], [56.89073, 11.78872], [56.84683, 11.81735], [56.84119, 11.821], [56.81912, 11.83457], [56.8072, 11.84163], [56.79812, 11.84673], [56.78627, 11.85305], [56.77885, 11.85673], [56.77203, 11.86003], [56.66897, 11.90792], [56.65783, 11.9129], [56.64872, 11.91673], [56.63902, 11.92055], [56.6284, 11.92518], [56.61781, 11.92958], [56.57957, 11.94565], [56.57033, 11.94953], [56.56059, 11.9536], [56.55168, 11.95727], [56.54181, 11.96133], [56.53172, 11.96552], [56.52187, 11.9696], [56.51215, 11.97368], [56.50232, 11.9779], [56.49295, 11.98178], [56.48315, 11.9859], [56.47343, 11.98997], [56.46368, 11.99408], [56.45027, 11.99977], [56.43801, 12.0049], [56.42828, 12.00892], [56.4178, 12.01327], [56.4085, 12.01717], [56.39821, 12.0213], [56.38743, 12.0257], [56.37659, 12.03032], [56.36457, 12.03535], [56.32705, 12.0508], [56.28725, 12.07742], [56.28243, 12.09083], [56.25593, 12.16377], [56.25105, 12.17738], [56.24625, 12.19095], [56.24165, 12.20402], [56.20825, 12.29282], [56.20299, 12.30627], [56.17624, 12.37192], [56.171, 12.385], [56.14303, 12.45367], [56.13839, 12.46517], [56.11477, 12.52225], [56.09503, 12.5636], [56.07501, 12.60465], [56.04995, 12.64698], [56.0418, 12.65232], [56.02725, 12.65715], [55.95039, 12.69775], [55.91233, 12.73387], [55.86803, 12.7425], [55.83547, 12.75332], [55.81118, 12.74025], [55.7988, 12.72782], [55.79095, 12.7197], [55.77917, 12.70768], [55.76903, 12.69717], [55.75897, 12.68722], [55.74947, 12.67747], [55.74078, 12.6681], [55.73553, 12.66215], [55.7274, 12.65278], [55.71723, 12.63505], [55.71388, 12.62762], [55.71125, 12.6207], [55.70862, 12.61337], [55.7055, 12.60702], [55.70403, 12.60283], [55.70403, 12.60283], [55.70418, 12.60268], [55.70082, 12.60125], [55.7002, 12.60103], [55.7002, 12.60098], [55.7002, 12.60102], [55.7002, 12.60103], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.7002, 12.601], [55.7002, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60098], [55.70022, 12.60102], [55.70022, 12.601], [55.7002, 12.60098], [55.7002, 12.60098], [55.7002, 12.60102], [55.7002, 12.601], [55.70022, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.70021, 12.60099], [55.7002, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60102], [55.70022, 12.601], [55.70022, 12.60102], [55.70022, 12.60102], [55.70022, 12.60098], [55.7002, 12.601], [55.70022, 12.60098], [55.7002, 12.60098], [55.7002, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70016, 12.60098], [55.7002, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60098], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60102], [55.70022, 12.601], [55.7002, 12.60098], [55.70022, 12.601], [55.7002, 12.60102], [55.70018, 12.60097], [55.7002, 12.60098], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.601], [55.70022, 12.60098], [55.7002, 12.60097], [55.70022, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.7002, 12.60098], [55.70022, 12.60098], [55.7002, 12.60098], [55.7002, 12.601], [55.70022, 12.601], [55.70022, 12.60098], [55.70022, 12.601], [55.7002, 12.60098], [55.7002, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.70022, 12.60098], [55.70022, 12.60098], [55.70022, 12.601], [55.70022, 12.601], [55.7002, 12.601], [55.70022, 12.601], [55.7002, 12.601], [55.7002, 12.60098], [55.7002, 12.601], [55.7002, 12.60098], [55.7002, 12.601]]

**Comment by annoviko** [comment]
@jordiarjona ,

Thank you for the provided data and description! I will investigate the problem.

**Comment by annoviko** [comment]
@jordiarjona ,

I get it :-) . Looks like you are trying to use medoids instead of medians. The principle difference between K-Medoids and K-Medians is that K-Medoids uses existed points from input data space as medoids, but median in K-Medians can be unreal object (not from input data space). So, if you need to use K-Medians algorithm, then you need to provide coordinates:
```python
    cluster_ids = random.sample(range(0, len(data)), 2)
    initial_medians = [ data[i] for i in cluster_ids ]

    kmedians_obj = kmedians(data, initial_medians, tolerance=tolerance, ccore=True)
    kmedians_obj.process()

    clusters = kmedians_obj.get_clusters();
    medians = kmedians_obj.get_medians();
```

In case of K-Medoids algorithm, you need to provide indexes from data as in your example:
```python
    tolerance = 10;
    cluster_ids = random.sample(range(0, len(data)), 2)
    initial_medians = [ data[i] for i in cluster_ids ]

    kmedoids_obj = kmedoids(data, cluster_ids, tolerance=tolerance, ccore=True)
    kmedoids_obj.process()

    clusters = kmedoids_obj.get_clusters();
    medoids = kmedoids_obj.get_medoids();
```

I guess it is good idea to add parameter checking to the algorithm constructor and print clear information about what is wrong :-) .

About tolerance: in line with provided data tolerance should be smaller, for example, 0.01 or 0.001. Otherwise only one clustering step is performed and result won't be accurate.

**Comment by jordiarjona** [comment]
Hi!

Sorry, my input was imprecise :)

I am doing what you also proposed, but i only pasted part of the code, my fault. The line where i define initial_centers was missing. I am doing the following:

cluster_ids = random.sample(range(0, len(data)), num_clusters)
initial_centers = [data[id] for id in cluster_ids]

before calling k-medians. I will test different tolerances, in case the problem is there.

**Comment by annoviko** [comment]
@jordiarjona ,

Yes, you are right, this is definetly bug in the code. Correction is the same for python part and C/C++ part, I will deliver correction. Thank you for the finding!

**Comment by annoviko** [comment]
@jordiarjona ,

Correction is delivered to 'master' and '0.8.dev' branches. And will be available in the next release 0.8.1. If you need changes now, here is an instruction how to install pyclustering from repository:
```bash
# download pyclustering from official repository:
mkdir pyclustering
cd pyclustering
git clone https://github.com/annoviko/pyclustering.git .

# compile C++ part of the library:
cd ccore
make ccore_x64   # in case of 64-bit python version
# make ccore_x86   # in case of 32-bit python version

# if you don't know which version of python is install then compile both
# make ccore 
```

In case Windows, C++ part of the library can be compiled by the following way:
1. Open 'pyclustering/ccore/' folder.
2. Open solution 'ccore.sln' using Microsoft Visual Studio 2015.
3. Choose platform (x64 - for 64-bit, x86 - for 32-bit).
4. Build 'ccore' project.
